File: common/seremis/soulcraft/soul/Soul.java
package seremis.soulcraft.soul;

import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;

public class Soul implements ISoul {

    protected IChromosome[] chromosomes;
    
    public Soul(IChromosome[] chomosomes) {
        this.chromosomes = chomosomes;
    }
    
    public Soul(NBTTagCompound compound) {
        readFromNBT(compound);
    }

    @Override
    public IChromosome[] getChromosomes() {
        return chromosomes;
    }

    @Override
    public void writeToNBT(NBTTagCompound compound) {
        compound.setInteger("genomeLength", chromosomes.length);
        
        NBTTagList tagList = new NBTTagList();
        
        for(IChromosome chromosome : chromosomes) {
            if(chromosome != null) {
                NBTTagCompound compound1 = new NBTTagCompound();
                chromosome.writeToNBT(compound1);
                tagList.appendTag(compound1);
            }
        }
        compound.setTag("chromosomes", tagList);
    }

    @Override
    public void readFromNBT(NBTTagCompound compound) {
        chromosomes = new Chromosome[compound.getInteger("genomeLength")];
        
        NBTTagList tagList = compound.getTagList("chromosomes");
        
        for(int i = 0; i < tagList.tagCount(); i++) {
            NBTTagCompound compound1 = (NBTTagCompound) tagList.tagAt(i);
            
            chromosomes[i] = new Chromosome(compound1);
        }
    }
}

File: src/main/scala/seremis/geninfusion/soul/Soul.scala
package seremis.geninfusion.soul

import net.minecraft.nbt.NBTTagCompound
import seremis.geninfusion.api.soul.{IChromosome, ISoul}
import seremis.geninfusion.api.util.AncestryNode
import seremis.geninfusion.util.GenomeHelper

class Soul(var chromosomes: Array[IChromosome], var name: Option[String] = None, var ancestry: AncestryNode) extends ISoul {

    if(ancestry != null)
        ancestry.chromosomes = chromosomes

    def this(chromosomes: Array[IChromosome], ancestry: AncestryNode) {
        this(chromosomes, None, ancestry)
    }
    
    def this(compound: NBTTagCompound) {
        this(null.asInstanceOf[Array[IChromosome]], None, null.asInstanceOf[AncestryNode])
        readFromNBT(compound)
        ancestry.chromosomes = chromosomes
    }

    override def getChromosomes: Array[IChromosome] = chromosomes
    override def getName: Option[String] = name
    override def getAncestryNode: AncestryNode = ancestry

    override def writeToNBT(compound: NBTTagCompound): NBTTagCompound = {
        name.foreach(n => compound.setString("soulName", n))
        
        compound.setTag("soulAncestry", ancestry.writeToNBT(new NBTTagCompound))

        compound
    }

    override def readFromNBT(compound: NBTTagCompound): NBTTagCompound = {
        if(compound.hasKey("soulName")) {
            name = Some(compound.getString("soulName"))
        }

        ancestry = AncestryNode.fromNBT(compound.getCompoundTag("soulAncestry"))
        chromosomes = ancestry.chromosomes

        chromosomes = GenomeHelper.fixGenomeErrors(this)

        compound
    }

    override def toString: String = {
        "Soul:[name: " + name + ", " + ancestry.toString + ", chromosomes:" + chromosomes.mkString(", ") + "]"
    }
}